Utils/Utils.hs

```diff
@@ -27,14 +28,32 @@ panic = error . (++) "Panic! Something went terribly wrong. "
 tack :: a -> [[a]] -> [[a]]
 tack x xs = (x : (head xs)) : (tail xs)
 
+partition :: PrettyPrint a => [(a -> Bool)] -> [a] -> [[a]]
+partition fs xs = go fs fs xs
+  where
+    go :: PrettyPrint a => [(a -> Bool)] -> [(a -> Bool)] -> [a] -> [[a]]
+    go _      _   []  = []
+    go (g:gs) gss (a:as)
+      | g a           = let (i, o) = part g (a:as) in (i : go gss gss o)
+      | otherwise     = go gs gss (a:as)
+    go _      _    as  = [panic ("No functions were provided." ++
+      " Or no function was able to match the content in 'partition'." ++
+      (render (ppr as)))]
+
+part :: (a -> Bool) -> [a] -> ([a], [a])
+part _ []     = ([], [])
+part f (x:xs)
+  | f x       = let (i, o) = part f xs in (x : i, o)
+  | otherwise = ([], (x:xs))
+
```

Comment:
I think you can use function `takeWhile` here to make the code more readable.

Reply:
I meant `span` actually (you need the remainder of the list as well):
```haskell
span :: (a -> Bool) -> [a] -> ([a], [a])
```

Reply:
Maybe something like this (I have not run this)?
```haskell
partition [] = []
partition ps@((isVar -> True):_) = varps : partition rest
  where  (varps, rest) = span isVar ps
partition ps@((isCon -> True):_) = conps : partition rest
  where  (conps, rest) = span isCon ps
partition ps@((isOr -> True):_) = orps : partition rest
  where  (orps, rest) = span isOr ps
partition ps = error $ "partition: impossible: " ++ render (ppr ps)
```

Reply:
Yeah just found `span` in hoogle as well. I think it would indeed be a lot neater if we didn't try to make partition so "general" because it was getting a little out of hand 😛

Reply:
Actually, we can avoid `ViewPatterns` altogether and go for the (even cleaner):
```haskell
partition [] = []
partition ps@(p:_)
  | isVar p, (varps, rest) <- span isVar ps = varps : partition rest
  | isCon p, (conps, rest) <- span isCon ps = conps : partition rest
  | isOr p , (orps, rest)  <- span isOr  ps = orps  : partition rest
partition ps = error $ "partition: impossible: " ++ render (ppr ps)
```